src/Analyzers/CSharp/Analyzers/ConvertTypeofToNameof/CSharpConvertTypeofToNameofDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertTypeofToNameof
+{
+    /// <summary>
+    /// Finds code like typeof(someType).Name and determines whether it can be changed to nameof(someType), if yes then it offers a diagnostic
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpConvertTypeofToNameofDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
```

Comment:
note: we would likely want thsi for VB as well (if the PR doesn't already support that).  i.e. `GetType(Foo).Name` shoudl be `nameof(Foo)`.  Because the analyzer is IOp based, it would/should be easy to create an impl that shares most of hte code between both with very little C#/VB specific behavior.